File: src/main/java/voidchess/player/ki/SimplePruner.java
package voidchess.player.ki;

/**
 * @author stephan
 */
public class SimplePruner implements SearchTreePruner {
    final static public int MAX_SEARCH_DEPTH = 6;
    final private int NO_HIT_RADIUS, FIGURE_HIT_RADIUS, CHESS_RADIUS;

    SimplePruner() {
        this(1, 2, 2);   //vergrößern dieser Variablen->bessere KI aber mehr Zeit nötig
    }

    public SimplePruner(int no_hit_radius, int figure_hit_radius, int chess_radius) {
        if (no_hit_radius < 1) throw new IllegalArgumentException("no_hit_radius has to be at least one");
        if (no_hit_radius > MAX_SEARCH_DEPTH) throw new IllegalArgumentException("no_hit_radius can't be bigger than MAX_SEARCH_DEPTH");
        if (figure_hit_radius < no_hit_radius)
            throw new IllegalArgumentException("figure_hit_radius cann't be smaller than no_hit_radius");
        if (figure_hit_radius > MAX_SEARCH_DEPTH) throw new IllegalArgumentException("figure_hit_radius can't be bigger than MAX_SEARCH_DEPTH");
        if (chess_radius < no_hit_radius)
            throw new IllegalArgumentException("chess_radius cann't be smaller than no_hit_radius");
        if (chess_radius > MAX_SEARCH_DEPTH) throw new IllegalArgumentException("chess_radius can't be bigger than MAX_SEARCH_DEPTH");

        NO_HIT_RADIUS = no_hit_radius;
        FIGURE_HIT_RADIUS = figure_hit_radius;
        CHESS_RADIUS = chess_radius;
    }

    public boolean stopMinDynamicEvaluation(int depth,
                                            boolean thisMove_isChess,
                                            boolean thisMove_hasHitFigure,
                                            boolean lastMove_isChess,
                                            boolean lastMove_hasHitFigure) {
        //return false;
        return !(depth < NO_HIT_RADIUS || thisMove_hasHitFigure || thisMove_isChess || lastMove_isChess);
    }

    public boolean stopMaxDynamicEvaluation(int depth,
                                            boolean thisMove_isChess,
                                            boolean thisMove_hasHitFigure,
                                            boolean lastMove_isChess,
                                            boolean lastMove_hasHitFigure) {
        if (depth < NO_HIT_RADIUS) return false;                                    // keine Verluste
        if (depth < FIGURE_HIT_RADIUS &&
                (thisMove_hasHitFigure)                                             // eine Figure wurde während des
        )
            return false;                                                           // letzten Halbzuges geschlagen
        if (depth < CHESS_RADIUS &&
                (thisMove_isChess || lastMove_isChess)                              // Schach gesetzt während des
        )
            return false;                                                           // letzten Halbzuges

        return true;
    }
}

File: src/main/kotlin/voidchess/player/ki/evaluation/SimplePruner.kt
package voidchess.player.ki.evaluation


open class SimplePruner(private val NO_HIT_RADIUS: Int = 1, private val FIGURE_HIT_RADIUS: Int = 2, private val CHESS_RADIUS: Int = 2) : SearchTreePruner {

    init {
        if (NO_HIT_RADIUS < 1) throw IllegalArgumentException("no_hit_radius has to be at least one")
        if (NO_HIT_RADIUS > MAX_SEARCH_DEPTH) throw IllegalArgumentException("no_hit_radius can't be bigger than MAX_SEARCH_DEPTH")
        if (FIGURE_HIT_RADIUS < NO_HIT_RADIUS)
            throw IllegalArgumentException("figure_hit_radius cann't be smaller than no_hit_radius")
        if (FIGURE_HIT_RADIUS > MAX_SEARCH_DEPTH) throw IllegalArgumentException("figure_hit_radius can't be bigger than MAX_SEARCH_DEPTH")
        if (CHESS_RADIUS < NO_HIT_RADIUS)
            throw IllegalArgumentException("chess_radius cann't be smaller than no_hit_radius")
        if (CHESS_RADIUS > MAX_SEARCH_DEPTH) throw IllegalArgumentException("chess_radius can't be bigger than MAX_SEARCH_DEPTH")
    }

    override fun stopMinDynamicEvaluation(depth: Int,
                                          thisMove_isChess: Boolean,
                                          thisMove_hasHitFigure: Boolean,
                                          lastMove_isChess: Boolean,
                                          lastMove_hasHitFigure: Boolean): Boolean {
        //return false;
        return !(depth < NO_HIT_RADIUS || thisMove_hasHitFigure || thisMove_isChess || lastMove_isChess)
    }

    override fun stopMaxDynamicEvaluation(depth: Int,
                                          thisMove_isChess: Boolean,
                                          thisMove_hasHitFigure: Boolean,
                                          lastMove_isChess: Boolean,
                                          lastMove_hasHitFigure: Boolean): Boolean {
        if (depth < NO_HIT_RADIUS) return false                                    // no figure was taken
        if (depth < FIGURE_HIT_RADIUS && thisMove_hasHitFigure) return false       // a figure was taken

        return !(depth < CHESS_RADIUS && (thisMove_isChess || lastMove_isChess))   // check now or a half move ago
    }

    companion object {
        const val MAX_SEARCH_DEPTH = 6
    }
}
